### DISCUSSION TITLE
Using the i2c extender plugin with the PCA9654E for matrix style button layouts

### DISCUSSION THREAD

**Discussion by ithinkido** [grblHAL/RP2040#33] [Q&A]
I would like to retrofit a machine that uses buttons that are arranged in a 3 x 8 matrix. I am not sure what would be the best way to add these extra io s to the PICO. One ideas was to use the PCA9654E plugin or something similar.  I am not sure if I understood correctly, but it looks this class only handles io pins in a linear array, and not in a matrix array? Is there a another way that I could look at to solve this issue . 
Something like [this ](https://github.com/joeyoung/arduino_keypads/tree/master/Keypad_I2Ca)based on a 16 bit io expanded chips would be ideal

**Comment by terjeio** [comment]
The [keypad plugin](https://github.com/grblHAL/Plugin_keypad/tree/85d8b70957516b5de7fe566fcbf6c83b88105d64) can read keycodes from an I2C device, it uses a [strobe signal](https://github.com/grblHAL/Plugin_keypad/blob/85d8b70957516b5de7fe566fcbf6c83b88105d64/keypad.c#L406-L424) to trigger a read so the interface is pretty simple.

There are I2C keypad handling chips available that should be usable with the interface, however I decided to use a cheap MSP430 MCU for [my implementation](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack/blob/master/Firmware/Keypad%20processor/main.c).

You may also add your own board specifc code that uses an I/O expander directly, but I am unsure if this requires constant polling of the expander to read the key status - that would depend on the capabilities of the expander? Polling an I2C device is, IMO, not desirable.

Without knowing what the keypad is going to be used for it is hard to tell what will be the best/easiest way to solve this issue.

**Reply by ithinkido** [reply]
Sorry for the late reply, I somehow missed your post. The keypad has functions for tool select ( 1 - 8 ) , basic x - y jog ( no speed nor distance options , only left - right , up - down ) and a few other non time critical functions similar to clamp hold and solenoid on / off. 
All of these are not time cricical, so I am not sure what would be the best solution , a seperate MCU or a i2c expander chip like the PCA9654E.  I think the i/o expander does have strobe , but I am still not sure if the [code](https://github.com/grblHAL/RP2040/blob/a6bf2c59b558eca057cca8a0f74566ed9ed75e78/ioexpand.h) is able to handle row + column arranged buttons and I am kind of stuck with that arrangement, as breaking out the buttons individually  is not easily possible.

**Reply by terjeio** [reply]
I don't know if it is possible to set up PCA9654E for a keypad matrix and to provide a strobe signal, you have to check the datasheet for that. 

The driver itself has an option for assigning an [I2C strobe pin](https://github.com/grblHAL/RP2040/blob/64a52c120286365b967c27a8a735ad7f2b844d9c/pico_cnc_map.h#L167-L170), and your code may claim the interrupt in the same way as the keypad plugin does it.

However, since you need jogging it might be easier for you to program a cheap MCU to scan the keypad, then you might even use the keypad plugin to handle some of the needed functionality. You can even add to that by trapping keypresses via the `keypad.on_keypress_preview` event...

**Reply by ithinkido** [reply]
Thanks for the advice. I will set up a separate controller to take over this task
